doc/designs/fips_indicator.md

```diff
@@ -0,0 +1,291 @@
+#else
+    return 0;
+#endif
+}
+~~~
+
+and in the algorithms get_ctx() function
+
+~~~
+int xxx_get_fips_approved(OSSL_PARAM params[])
+{
+    p = OSSL_PARAM_locate(params, OSSL_FIPS_PARAM_APPROVED);
+    if (p != NULL && !OSSL_PARAM_set_int(p, ossl_xxx_fips_approved()))
+        return 0;
+	return 1;
+}
+~~~
+
+#### API’s that would be used to support this are
+
+- EVP_PKEY Keygen, Encryption, Signatures, Key Exchange, KEM
+~~~ 
+EVP_PKEY_CTX_get_params(ctx, );
+~~~
+(Note that this would mean you couldnt use functions that hide the ctx such as EVP_PKEY_Q_keygen()!)
+
+- Ciphers
+~~~
+EVP_CIPHER_CTX_get_params()
+~~~
+
+- Digests
+~~~
+EVP_MD_CTX_get_params()
+~~~
+
+- KDF’s
+~~~
+EVP_KDF_CTX_get_params()
+~~~
+
+- MAC’s
+~~~
+EVP_MAC_CTX_get_params()
+~~~
+
+- RAND
+~~~
+EVP_RAND_CTX_get_params()
+~~~
+
+#### Backwards Compatability..
+
+Previous providers do not support this operation, so they will return not approved if they are not handled.
+
+#### Alternate Solution
+
+If we had different kinds of compliance requirements (something other than FIPS) either a seperate getter could be added or the getter could return a int type instead of a 0 or 1..
+(e.g 1 = fips approved, 2 = some other compliance approved)
+
+
+## Changes Required for indicators
+
+### key size >= 112 bits
+
+There are a few places where we do not enforce key size that need to be addressed.
+
+- HMAC
+Which applies to all algorithms that use HMAC also.
+- CMAC
+
+### Algorithm Transitions
+
+Should we remove these algorithms completely from the fips provider, or use indicators? 
+
+- DES_EDE3_ECB
+Disallowed for encryption, allowed for legacy decryption
+
+- DSA
+Keygen and Signing are no longer approved, not sure if verify is still approved.
+
+- RSA Signing using PKCSV15
+
+Rsa self test for sign may need to change as the default pad_mode is PKCSV15
+Check if saltlen needs a indicator
+Padding mode updates required in rsa_check_padding().
+Check if sha1 is allowed?
+
+- ECDSA B & K curves are deprecated
+It looks like these are still allowed. Are the approved?
+
+- ED25519/ED448 is now approved
+
+- X25519/X448 is not approved currently. keygen would need an indicator if we allow it?
+
+- RSA Key transport
+Padding need to be changed to either not allow padding modes other than OEAP or use an indicator.
+
+- RSA
+It looks like SP800-131Ar2 specifies that RSA >= 2048 is approved.
+Can we check with the labs what to do for the legacy verification cases
+since we allow >=1024. Would this now be unapproved?
+This would apply to rsa_kem also?
+rsa_keygen_pairwise_test() may need to change to do a sign/verify PCT?
+
+- TLS1_PRF
+If we are only trying to support TLS1.2 here then we should remove the tls1.0/1.1 code from the FIPS MODULE.
+
+
+### Digest Checks
+
+Any algorithms that use a digest need to make sure that the CAVP certificate lists all supported FIPS digests otherwise an indicator is required.
```

Comment:
Currently Hash and HMAC DRBGs are the only impacted algorithms.
Nonetheless, this is a concern going forwards.

Reply:
I disagree. The DRBG's are the only algorithms that are doing the right thing currently because you added code to check what digests are allowed. Unless the FIPS cert lists all the digests that matches all the FIPS digest algorithms then we are not currently doing it correctly for a FIPS 140-3 validation.

Reply:
The 140-3 should list everything.  I've not seen the actual submission so I don't know if this is correct or not.

Reply:
I find it hard to believe it was done for absolutely everything.. (including things like SSKDF, and HKDF)

Reply:
I am not sure I understand your point, look at the 3.0.8 validation for example and you will see that the digests tested vary wildly. This could be because there was not even an option to test them at that point in time, I dont know since I wasnt part of that process.

Reply:
The commit message in https://gitlab.com/redhat/centos-stream/rpms/openssl/-/blob/c9s/0078-KDF-Add-FIPS-indicators.patch?ref_type=heads has a nice overview of which combinations can be tested, and which ones we were told to consider unapproved (and thus marked with an indicator saying so).